File: ai-poller/encryption.py
"""AES-256-GCM encryption compatible with WorkPlanner web and Android apps.

Key derivation: PBKDF2-HMAC-SHA256, 210,000 iterations, 16-byte salt -> 32-byte key.
Encrypt format: [IV (12 bytes)][ciphertext + auth tag (16 bytes)].
"""

import os

from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
from cryptography.hazmat.primitives import hashes

PBKDF2_ITERATIONS = 210_000
KEY_LENGTH_BYTES = 32
SALT_LENGTH_BYTES = 16
IV_LENGTH_BYTES = 12


def generate_salt() -> bytes:
    return os.urandom(SALT_LENGTH_BYTES)


def derive_key(passphrase: str, salt: bytes) -> bytes:
    kdf = PBKDF2HMAC(
        algorithm=hashes.SHA256(),
        length=KEY_LENGTH_BYTES,
        salt=salt,
        iterations=PBKDF2_ITERATIONS,
    )
    return kdf.derive(passphrase.encode("utf-8"))


def encrypt(plain_bytes: bytes, key: bytes) -> bytes:
    iv = os.urandom(IV_LENGTH_BYTES)
    aesgcm = AESGCM(key)
    ciphertext_and_tag = aesgcm.encrypt(iv, plain_bytes, None)
    return iv + ciphertext_and_tag


def decrypt(encrypted_bytes: bytes, key: bytes) -> bytes:
    iv = encrypted_bytes[:IV_LENGTH_BYTES]
    ciphertext_and_tag = encrypted_bytes[IV_LENGTH_BYTES:]
    aesgcm = AESGCM(key)
    return aesgcm.decrypt(iv, ciphertext_and_tag, None)

File: web/src/crypto/encryption.ts
const PBKDF2_ITERATIONS = 210_000;
const KEY_LENGTH_BITS = 256;
const SALT_LENGTH_BYTES = 16;
const IV_LENGTH_BYTES = 12;
const TAG_LENGTH_BITS = 128;

export function generateSalt(): Uint8Array {
  return crypto.getRandomValues(new Uint8Array(SALT_LENGTH_BYTES));
}

export async function deriveKey(
  passphrase: string,
  salt: Uint8Array,
): Promise<Uint8Array> {
  const encoder = new TextEncoder();
  const keyMaterial = await crypto.subtle.importKey(
    'raw',
    encoder.encode(passphrase),
    'PBKDF2',
    false,
    ['deriveBits'],
  );

  const derivedBits = await crypto.subtle.deriveBits(
    {
      name: 'PBKDF2',
      salt: salt.buffer as ArrayBuffer,
      iterations: PBKDF2_ITERATIONS,
      hash: 'SHA-256',
    },
    keyMaterial,
    KEY_LENGTH_BITS,
  );

  return new Uint8Array(derivedBits);
}

export async function encrypt(
  plainBytes: Uint8Array,
  keyBytes: Uint8Array,
): Promise<Uint8Array> {
  const iv = crypto.getRandomValues(new Uint8Array(IV_LENGTH_BYTES));

  const cryptoKey = await crypto.subtle.importKey(
    'raw',
    keyBytes.buffer as ArrayBuffer,
    'AES-GCM',
    false,
    ['encrypt'],
  );

  const ciphertext = await crypto.subtle.encrypt(
    { name: 'AES-GCM', iv, tagLength: TAG_LENGTH_BITS },
    cryptoKey,
    plainBytes.buffer as ArrayBuffer,
  );

  // Prepend IV to ciphertext (which includes the auth tag)
  const result = new Uint8Array(IV_LENGTH_BYTES + ciphertext.byteLength);
  result.set(iv, 0);
  result.set(new Uint8Array(ciphertext), IV_LENGTH_BYTES);
  return result;
}

export async function decrypt(
  encryptedBytes: Uint8Array,
  keyBytes: Uint8Array,
): Promise<Uint8Array> {
  const iv = encryptedBytes.slice(0, IV_LENGTH_BYTES);
  const ciphertext = encryptedBytes.slice(IV_LENGTH_BYTES);

  const cryptoKey = await crypto.subtle.importKey(
    'raw',
    keyBytes.buffer as ArrayBuffer,
    'AES-GCM',
    false,
    ['decrypt'],
  );

  const plainBuffer = await crypto.subtle.decrypt(
    { name: 'AES-GCM', iv, tagLength: TAG_LENGTH_BITS },
    cryptoKey,
    ciphertext.buffer as ArrayBuffer,
  );

  return new Uint8Array(plainBuffer);
}
